src/Quasar.jl

```diff
@@ -0,0 +1,1917 @@
+    elseif head(expr) == :array_literal
+        return [evaluate(v, arg) for arg in convert(Vector{QasmExpression}, expr.args)]
+    elseif head(expr) == :observable
+        raw_obs = expr.args[1]::QasmExpression
+        if head(raw_obs) == :array_literal
+            new_obs = map(arg->evaluate(v, QasmExpression(:observable, arg)), convert(Vector{QasmExpression}, raw_obs.args))
+            return Braket.Observables.TensorProduct(convert(Vector{Braket.Observables.Observable}, new_obs))
+        elseif head(raw_obs) == :identifier
+            obs_name = raw_obs.args[1]::String
+            obs_name == "x" && return Braket.Observables.X()
+            obs_name == "y" && return Braket.Observables.Y()
+            obs_name == "z" && return Braket.Observables.Z()
+            obs_name == "i" && return Braket.Observables.I()
+            obs_name == "h" && return Braket.Observables.H()
+        elseif head(raw_obs) == :hermitian
+            h_mat = similar(raw_obs.args[1], ComplexF64)::Matrix{ComplexF64}
+            for ii in eachindex(h_mat)
+                h_mat[ii] = convert(ComplexF64, evaluate(v, raw_obs.args[1][ii]))::ComplexF64
+            end
+            return Braket.Observables.HermitianObservable(h_mat)
+        end
+    elseif head(expr) == :power_mod
+        pow_expr = QasmExpression(:pow, evaluate(v, expr.args[1]::QasmExpression))
+        return (pow_expr, expr.args[2])
+    elseif head(expr) == :inverse_mod
+        return (QasmExpression(:inv), expr.args[1]::QasmExpression)
+    elseif head(expr) ∈ (:control_mod, :negctrl_mod)
+        has_argument = length(expr.args) > 1
+        if has_argument
+            arg_val::Int = evaluate(v, first(expr.args)::QasmExpression)::Int
+            isinteger(arg_val) || throw(QasmVisitorError("cannot apply non-integer ($arg_val) number of controls or negcontrols."))
+            true_inner = expr.args[2]::QasmExpression
+            inner = QasmExpression(head(expr), true_inner)
+            while arg_val > 2
+                inner = QasmExpression(head(expr), inner)
+                arg_val -= 1
+            end
+        else
+            inner = expr.args[1]::QasmExpression
+        end
+        new_head = head(expr) == :control_mod ? :ctrl : :negctrl
+        return (QasmExpression(new_head), inner)
+    elseif head(expr) == :binary_op
+        op  = expr.args[1]::Symbol
+        lhs = evaluate(v, expr.args[2])
+        rhs = evaluate(v, expr.args[3])
+        val = evaluate_binary_op(op, lhs, rhs)
+        return val 
+    elseif head(expr) == :unary_op
+        op  = expr.args[1]::Symbol
+        arg = evaluate(v, expr.args[2])
+        return evaluate_unary_op(op, arg)
+    elseif head(expr) == :hw_qubit
+        return tryparse(Int, replace(expr.args[1], "\$"=>""))
+    elseif head(expr) == :cast
+        casting_to = expr.args[1].args[1]
+        value = evaluate(v, expr.args[2])
+        if casting_to == Bool
+            return value > 0
+        # TODO
```

Comment:
What's this TODO for?

Reply:
Casting to non-bool values explicitly. Currently none of our tests or examples are doing this.